feat: add HTTP hook Secret Configuration

J0 opened the pull request:
## What kind of change does this PR introduce?

Adds the corresponding configuration for the `SendSMS` and `SendEmail` HTTP Hooks. Also introduces Secrets for previous hooks - `custom_access_token`, `mfa_verification_attempt` , `password_verification_attempt`


Backend PR (Merged): https://github.com/supabase/infrastructure/pull/17848
Frontend PR (WIP): https://github.com/supabase/supabase/pull/22495

J0 commented on internal/utils/config.go:
```diff
@@ -674,6 +677,29 @@ func LoadConfigFS(fsys afero.Fs) error {
+					return errors.New("Missing required field in config: auth.hook.send_email.uri")
+				}
+				if !strings.HasPrefix(strings.ToLower(Config.Auth.Hook.SendEmail.URI), "http") {
+					return errors.New("Invalid HTTP hook config: auth.hook.send_email.uri should be a http or https URL")
```
```suggestion
					return errors.New("Invalid HTTP hook config: auth.hook.send_email.uri should be a HTTP or HTTPS URL")
```

J0 commented on internal/utils/misc.go:
```diff
@@ -40,7 +40,7 @@ const (
 	PgbouncerImage   = "bitnami/pgbouncer:1.20.1-debian-11-r39"
 	PgProveImage     = "supabase/pg_prove:3.36"
-	GotrueImage      = "supabase/gotrue:v2.145.0"
+	GotrueImage      = "supabase/gotrue:v2.147.1"
```
This isn't out yet, but will be deployed soon. Currently, we are on v2.147.0 on middleware and v2.146 for all projects. Can wait to bump this too if that's better.

J0 replied:
```suggestion
	GotrueImage      = "supabase/gotrue:v2.148.0"
```

sweatybridge commented on internal/utils/config.go:
```diff
@@ -674,6 +677,29 @@ func LoadConfigFS(fsys afero.Fs) error {
+				if Config.Auth.Hook.SendSMS.URI == "" {
+					return errors.New("Missing required field in config: auth.hook.send_sms.uri")
+				}
+				if !strings.HasPrefix(strings.ToLower(Config.Auth.Hook.SendSMS.URI), "http") {
```
Shall we parse the URL and check for scheme instead?

sweatybridge commented on internal/utils/templates/init_config.toml:
```diff
@@ -119,6 +119,20 @@ template = "Your code is {{ `{{ .Code }}` }} ."
+[auth.hook.send_sms]
+# enabled = true
+# We use host.docker.internal to connect to port 54321 on the host. See: https://docs.docker.com/desktop/networking/
+# uri = "http://host.docker.internal:54321/functions/v1/sms_sender"
```
Since this hook is called from inside auth container, it should be possible to use either `kong` or `edge_runtime` network alias, ie. `http://kong:8000/functions/v1/sms_sender`

J0 replied:
Sure let's use `kong`  so it goes through the gateway still.

J0 replied:
Ah sigh. This would require an auth level change

J0 replied:
Opened a pr, but for reference do you happen to know which platforms?

sweatybridge replied:
Linux was known to have problems https://stackoverflow.com/a/62431165 so we added the default workaround with `host-gateway`. However the workaround doesn't work for older versions of podman https://github.com/containers/podman/issues/14390.

But it appears that all these issues are resolved now so may be it's ok. 😄

sweatybridge commented on internal/utils/templates/init_config.toml:
```diff
@@ -119,6 +119,20 @@ template = "Your code is {{ `{{ .Code }}` }} ."
+# Use a custom sms sender
+[auth.hook.send_sms]
+# enabled = true
+# We use host.docker.internal to connect to port 54321 on the host. See: https://docs.docker.com/desktop/networking/
```
This comment looks outdated.

sweatybridge commented on internal/utils/templates/init_config.toml:
```diff
@@ -119,6 +119,20 @@ template = "Your code is {{ `{{ .Code }}` }} ."
+# We use host.docker.internal to connect to port 54321 on the host. See: https://docs.docker.com/desktop/networking/
+# uri = "http://kong:8000/functions/v1/sms_sender"
+# Use this secret in the email_sender edge function as well.
+# secrets = "v1,whsec_c3VwYWJhc2Vpc3Zlcnlhd2Vzb21lYW5kaWxpa2VpdA=="
```
How is this secret generated? If it's only a static value, may be we can hardcode it in config.go. Otherwise we want to use `env(SOME_SECRET)` substitution.

To avoid accidents and confusion, we don't want to add secret to config.toml even if they are only used locally.

J0 commented:
Actually I think we’ll need to wait for the FE to go out before merging :/

sweatybridge commented on internal/utils/config.go:
```diff
@@ -357,11 +358,14 @@ type (
-		URI     string `toml:"uri"`
+		Enabled         bool   `toml:"enabled"`
+		URI             string `toml:"uri"`
+		HTTPHookSecrets string `toml:"secrets"`
```
```suggestion
		Secrets string `toml:"secrets"`
```

Shall we just name this field `Secrets` for consistency with toml tag?

J0 replied:
Yup. I’ll need to add an additional set of config vars with our latest changes as well.

J0 commented:
New Changes:
- Removed the `init_config.toml` 
- Added tests 

The backend is now merged and I'm hoping to get this in so that we can allow [contributors to write examples](https://github.com/supabase/auth/pull/1018#issuecomment-2098612913)

Tested locally

sweatybridge commented on internal/utils/config.go:
```diff
@@ -460,6 +464,24 @@ type (
+		return nil
+	}
+	if h.URI == "" {
+		return fmt.Errorf("missing required field in config: auth.hook.%s.uri", hookType)
```
```suggestion
		return errors.Errorf("missing required field in config: auth.hook.%s.uri", hookType)
```

use `go-errors` package for auto stack trace.

J0 replied:
Thanks! TIL

sweatybridge commented on internal/utils/config.go:
```diff
@@ -460,6 +464,24 @@ type (
+	}
+	var err error
+	if h.Secrets, err = maybeLoadEnv(h.Secrets); err != nil {
+		return fmt.Errorf("missing required field in config: auth.hook.%s.secrets", hookType)
```
```suggestion
		return errors.Errorf("missing required field in config: auth.hook.%s.secrets", hookType)
```

ditto

sweatybridge reviewed (approved):
Might want to add a valid config to `init_config.test.toml` as a sanity check.
